erts/emulator/beam/erl_arith.c

```diff
@@ -461,6 +461,70 @@ erts_mixed_plus(Process* p, Eterm arg1, Eterm arg2)
     }
 }
 
+Eterm
+erts_unary_minus(Process* p, Eterm arg)
+{
+    DECLARE_TMP(tmp_big,0,p);
+    Eterm hdr, res, zero;
+    FloatDef f;
+    dsize_t sz1, sz2, sz;
+    int need_heap;
+    Eterm* hp;
+    Sint ires;
+
+    ERTS_FP_CHECK_INIT(p);
+    switch (arg & _TAG_PRIMARY_MASK) {
+    case TAG_PRIMARY_IMMED1:
+        switch ((arg & _TAG_IMMED1_MASK) >> _TAG_PRIMARY_SIZE) {
+        case (_TAG_IMMED1_SMALL >> _TAG_PRIMARY_SIZE):
+            ires = -signed_val(arg);
+            if (IS_SSMALL(ires)) {
+                return make_small(ires);
+            } else {
+                hp = HeapFragOnlyAlloc(p, 2);
+                res = small_to_big(ires, hp);
+                return res;
+            }
+        default:
+        badarith:
+            p->freason = BADARITH;
+            return THE_NON_VALUE;
+        }
+    case TAG_PRIMARY_BOXED:
+        hdr = *boxed_val(arg);
+        switch ((hdr & _TAG_HEADER_MASK) >> _TAG_PRIMARY_SIZE) {
+        case (_TAG_HEADER_POS_BIG >> _TAG_PRIMARY_SIZE):
+        case (_TAG_HEADER_NEG_BIG >> _TAG_PRIMARY_SIZE):
+            zero = small_to_big(signed_val(make_small(0)), tmp_big);
+            sz1 = big_size(zero);
+            sz2 = big_size(arg);
+            sz = MAX(sz1, sz2)+1;
+            need_heap = BIG_NEED_SIZE(sz);
+            hp = HeapFragOnlyAlloc(p, need_heap);
+            res = big_minus(zero, arg, hp);
+            maybe_shrink(p, hp, res, need_heap);
+            if (is_nil(res)) {
+                p->freason = SYSTEM_LIMIT;
+                return THE_NON_VALUE;
+            }
+            return res;
```

Comment:
It feels there is a more efficient version of this whole branch but I am well versed enough to come up with it.

Reply:
Big numbers have explicit sign bits (just as floats). The sign bit is in the header word (_TAG_HEADER_POS_BIG and _TAG_HEADER_NEG_BIG). There is indeed only a single bits difference (NEG_BIG_SUBTAG - POS_BIG_SUBTAG). You can implement the negation by simply copying the original bignum, followed by a sign flip of the header word. Be sure to add a call to ERTS_CT_ASSERT() to ensure that the bit flipping is safe, for example (not tested):

```
ERTS_CT_ASSERT((NEG_BIG_SUBTAG - POS_BIG_SUBTAG) >> _TAG_PRIMARY_SIZE == 1)
```

Reply:
Thanks @bjorng!

Unfortunately I tried this and my C knowledge is not good enough to make this happen. In case you or anyone else wants to pick this up, it is totally fine by me. Sorry!